File: bada_ruby/lib/bada/quantum/semiconductor.rb
# frozen_string_literal: true

module Bada
  module Quantum
    # The semiconductor realisability principle (半導体で使える原理).
    #
    # For the telegraph to be buildable, the entangled pair must come from
    # something you can fabricate: a semiconductor quantum dot emitting a
    # polarization-entangled photon pair via the biexciton–exciton cascade, with
    # spin qubits read through the band structure. Two textbook relations decide
    # whether that hardware works at a given temperature:
    #
    #   * Fermi–Dirac occupation  f(E) = 1 / (1 + exp((E-μ)/kT))  — thermal
    #     population of carrier states; thermal excitation across the gap is the
    #     dominant qubit error source.
    #   * WKB tunnelling  T ≈ exp(-2 κ d),  κ = sqrt(2 m ΔE)/ħ  — how well a
    #     read/coupling barrier passes the carrier.
    #
    # We work in electron-volts with kT in eV so no SI juggling is needed, and
    # turn the thermal excitation into the channel's bit-flip probability.
    module Semiconductor
      module_function

      KB_EV = 8.617333262e-5 # Boltzmann constant in eV/K

      # A few representative direct-gap materials (eV) used for entangled-photon
      # sources / spin qubits.
      BAND_GAPS = {
        "GaAs"   => 1.42,
        "InAs"   => 0.354,
        "InGaAs" => 0.75,
        "Si"     => 1.12,
        "diamond(NV)" => 5.47
      }.freeze

      def kT(temp_k)
        KB_EV * temp_k
      end

      # Fermi–Dirac occupation of a state at energy E (eV) with chemical
      # potential mu (eV) at temperature T (K).
      def fermi_dirac(energy_ev, mu_ev, temp_k)
        x = (energy_ev - mu_ev) / kT(temp_k)
        x = 700.0 if x > 700.0 # avoid overflow
        x = -700.0 if x < -700.0
        1.0 / (1.0 + Math.exp(x))
      end

      # Thermal excitation probability across the gap (μ at mid-gap): the qubit
      # bit-flip probability from the lattice bath.
      def thermal_flip(material, temp_k)
        eg = BAND_GAPS.fetch(material)
        # occupation of the conduction edge with mu at mid-gap
        fermi_dirac(eg, eg / 2.0, temp_k)
      end

      # WKB tunnelling transmission through a barrier of height ΔE (eV) and
      # width d (nm) for an electron (m = m_e).
      def tunnelling(barrier_ev, width_nm)
        # κ [1/nm] = sqrt(2 m ΔE)/ħ ; with ΔE in eV this is ≈ 5.123 * sqrt(ΔE)
        kappa = 5.123 * Math.sqrt([barrier_ev, 0.0].max)
        Math.exp(-2.0 * kappa * width_nm)
      end

      # Feasibility report for a chosen material at operating temperature.
      def report(material: "GaAs", temp_k: 4.0, barrier_ev: 0.3, width_nm: 5.0)
        flip = thermal_flip(material, temp_k)
        {
          material: material,
          band_gap_ev: BAND_GAPS.fetch(material),
          temperature_k: temp_k,
          kT_ev: kT(temp_k),
          bit_flip_prob: flip,
          tunnelling: tunnelling(barrier_ev, width_nm),
          # a channel is "buildable" if thermal errors are correctable (<10%)
          feasible: flip < 0.1
        }
      end
    end
  end
end

File: apps/core/src/main/java/bada/quantum/Semiconductor.java
package bada.quantum;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Semiconductor realisability (半導体で使える原理): a GaAs quantum dot emitting
 * a polarization-entangled photon pair. Fermi–Dirac occupation gives the
 * thermal bit-flip probability; WKB tunnelling gives barrier transmission.
 * Energies in eV so kT stays in eV.
 */
public final class Semiconductor {
    public static final double KB_EV = 8.617333262e-5; // eV/K

    public static final Map<String, Double> BAND_GAPS = new LinkedHashMap<>();
    static {
        BAND_GAPS.put("GaAs", 1.42);
        BAND_GAPS.put("InAs", 0.354);
        BAND_GAPS.put("InGaAs", 0.75);
        BAND_GAPS.put("Si", 1.12);
        BAND_GAPS.put("diamond(NV)", 5.47);
    }

    public static double kT(double tempK) { return KB_EV * tempK; }

    public static double fermiDirac(double energyEv, double muEv, double tempK) {
        double x = (energyEv - muEv) / kT(tempK);
        if (x > 700) x = 700;
        if (x < -700) x = -700;
        return 1.0 / (1.0 + Math.exp(x));
    }

    /** Thermal excitation across the gap (mu at mid-gap) = qubit bit-flip prob. */
    public static double thermalFlip(String material, double tempK) {
        double eg = BAND_GAPS.getOrDefault(material, 1.42);
        return fermiDirac(eg, eg / 2.0, tempK);
    }

    /** WKB transmission through barrier height ΔE (eV), width d (nm). */
    public static double tunnelling(double barrierEv, double widthNm) {
        double kappa = 5.123 * Math.sqrt(Math.max(barrierEv, 0.0));
        return Math.exp(-2.0 * kappa * widthNm);
    }

    public static final class Report {
        public String material;
        public double bandGapEv, temperatureK, kTev, bitFlipProb, tunnelling;
        public boolean feasible;
    }

    public static Report report(String material, double tempK, double barrierEv, double widthNm) {
        Report r = new Report();
        r.material = material;
        r.bandGapEv = BAND_GAPS.getOrDefault(material, 1.42);
        r.temperatureK = tempK;
        r.kTev = kT(tempK);
        r.bitFlipProb = thermalFlip(material, tempK);
        r.tunnelling = tunnelling(barrierEv, widthNm);
        r.feasible = r.bitFlipProb < 0.1;
        return r;
    }

    public static Report report(String material, double tempK) {
        return report(material, tempK, 0.3, 5.0);
    }
}
